bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairBaseBridgeHandler.java

```diff
@@ -99,355 +143,317 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-            byte[] buf = new byte[5 + p.getLength() + 8]; // 5 is preamble, 8 is 6 bytes for header and 2 for checksum
+        writePacket(p, -1, 0);
+    }
+
+    public boolean writePacket(PentairPacket p, int response, int retries) {
+        boolean bReturn = true;
 
+        try {
+            byte[] buf;
+            int nRetries = retries;
+
+            if (!writer.isPresent()) {
+                logger.debug("writePacket: writer = null");
+                return false;
+            }
             p.setSource(id);
 
-            System.arraycopy(preamble, 0, buf, 0, 5);
-            System.arraycopy(p.buf, 0, buf, 5, p.getLength() + 6);
-            int checksum = p.calcChecksum();
+            buf = p.getFullWriteStream();
 
-            buf[p.getLength() + 11] = (byte) ((checksum >> 8) & 0xFF);
-            buf[p.getLength() + 12] = (byte) (checksum & 0xFF);
+            lock.lock();
+            ackResponse = response;
 
-            logger.debug("Writing packet: {}", PentairPacket.bytesToHex(buf));
+            do {
+                logger.debug("Writing packet: {}", PentairPacket.toHexString(buf));
 
-            writer.write(buf, 0, 5 + p.getLength() + 8);
-            writer.flush();
+                writer.get().write(buf, 0, buf.length);
+                writer.get().flush();
+
+                if (response != -1) {
+                    logger.debug("writePacket: wait for ack (response: {}, retries: {})", response, nRetries);
+                    bReturn = waitAck.await(1000, TimeUnit.MILLISECONDS); // bReturn will be false if timeout
+                    nRetries--;
+                }
+            } while (!bReturn && (nRetries >= 0));
         } catch (IOException e) {
-            logger.trace("I/O error while writing stream", e);
+            logger.debug("I/O error while writing stream: {}", e.getMessage());
+            internalDisconnect();
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (InterruptedException e) {
+            logger.debug("writePacket: InterruptedException: {}", e.getMessage());
+            Thread.currentThread().interrupt();
+        } finally {
+            lock.unlock();
+        }
+
+        if (!bReturn) {
+            logger.debug("writePacket: timeout");
+        }
+        return bReturn;
+    }
+
+    /**
+     * Method to acknowledge an ack or response packet has been sent
+     *
+     * @param cmdresponse is the command that was seen as a return. This is validate against that this was the response
+     *            before signally a return.
+     */
+    public void ackResponse(int response) {
+        if (response != ackResponse) {
+            return;
+        }
+
+        try {
+            lock.lock();
+            waitAck.signalAll();
+            lock.unlock();
+        } catch (IllegalMonitorStateException e) {
+            logger.debug("ackResponse: illegal monitor exception: {}", e.getMessage());
```

Comment:
How is this getting thrown?

Reply:
from the signalAll documentation:

void signalAll()
Wakes up all waiting threads.
If any threads are waiting on this condition then they are all woken up. Each thread must re-acquire the lock before it can return from await.

Implementation Considerations

An implementation may (and typically does) require that the current thread hold the lock associated with this Condition when this method is called. Implementations must document this precondition and any actions taken if the lock is not held. Typically, an exception such as IllegalMonitorStateException will be thrown.

Reply:
What I was trying to point out was that you are already doing proper locking before you signal so there there should be no way for this exception to get thrown. It is safe to remove the catch for this exception.

Reply:
Sounds good.  was just trying to make sure all conditions were covered.